### ISSUE TITLE
NativeAot Windows DLL import leftovers

### ISSUE THREAD

**Issue by yowl** [dotnet/runtimelab#490]
Continuing from https://github.com/dotnet/corert/issues/8254.  I still have some of these getting into the Ilc output for Wasm:
```
EXEC : warning : undefined symbol: CompareStringEx (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: CompareStringOrdinal (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: FindNLSStringEx (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: FindStringOrdinal (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: GetCalendarInfoEx (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: GetLocaleInfoEx (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: LCMapStringEx (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
EXEC : warning : undefined symbol: LocaleNameToLCID (referenced by top-level compiled C/C++ code) [E:\GitHub\runtimelab\samples\HelloWorld\HelloWorld.csproj]
```

Taking `CompareStringEx` it is not eliminated by the dependencies as its brought in by (mangled names), from lower to higher in the dependency call graph

`S_P_CoreLib_System_Globalization_CompareInfo__NlsCompareString`
`S_P_CoreLib_System_Globalization_CompareInfo__CompareStringCore`
`S_P_CoreLib_System_Globalization_CompareInfo__Compare_5`

At runtime maybe this path cant be reached, but if you try to start a Wasm program in wasmer (https://wasmer.io/), unlike the Wasm VMs in browsers, it refuses to start if there are undefined imports:
```
PS C:\WINDOWS\system32> wasmer E:\GitHub\runtimelab\samples\HelloWorld\bin\wasm\Debug\net5.0\browser-wasm\publish\HelloWorld.wasm
error: failed to run `E:\GitHub\runtimelab\samples\HelloWorld\bin\wasm\Debug\net5.0\browser-wasm\publish\HelloWorld.wasm`
╰─> 1: Error while importing "env"."CompareStringEx": unknown import. Expected Function(FunctionType { params: [I32, I32, I32, I32, I32, I32, I32, I32, I32], results: [I32] })
```

Would it be appropriate for `TARGET_WASM` to `#ifdef` them out somewhere in the above list, or maybe provide a stub for `S_P_CoreLib_Interop_Kernel32__CompareStringEx` and the others?

**Comment by MichalStrehovsky** [comment]
That's odd. I wouldn't expect references to those APIs to exist outside Windows. It's weird they're not in `.Win32.cs` files. I guess on CoreCLR this calls into the Win32 API emulator that is the CoreCLR "PAL". Not sure what that does on Mono (or CoreRT for that matter) outside Windows.

Seems like IL Linker kills it off during Mono's WASM CoreLib build (`build.cmd mono.corelib -os Browser -arch Wasm` -`CompareStringCore` has weird IL because IL Linker removed the "use NLS" branch at the time of build). We could just ifdef that out for WASM, but since Mono is not doing that and to reduce the diff to dotnet/runtime, we can just do the same thing mono does.

I just added the IL Linker feature that is needed to our compiler in #422. We don't run IL linker at library build time because it would kill off stuff the compiler needs and I can't be bothered to teach linker not to do that - we wouldn't benefit much from IL Linker in general since the compiler does that out of box.

What would probably work:

* Steal Mono's WASM substitutions file (or link to it): https://github.com/dotnet/runtime/blob/1e2685b6abaeb38013bcd7090db5301f14f725f8/src/mono/netcore/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.wasm.xml
* Add a reference to the file from CoreLib's csproj, like mono does here: https://github.com/dotnet/runtime/blob/1e2685b6abaeb38013bcd7090db5301f14f725f8/src/mono/netcore/System.Private.CoreLib/System.Private.CoreLib.csproj#L135-L136
* Add an extra entry to the substitution file that hardcodes `GlobalizationMode.get_UseNls` to `false`

This should result in the compiler stripping away the codepaths that do if/else on this but are not reachable at runtime.

**Comment by jkotas** [comment]
The substitution file that sets UseNls to false on non-Windows should be a good idea in NativeAOT branch too.

**Comment by yowl** [comment]
Ok, I will try the `UseNls` first in NativeAOT, and see if I can get the symbol to disappear from the final executable in Linux.

**Comment by yowl** [comment]
Having a small problem running the tests on Linux/Ubuntu.  

` src/tests/run.sh --runnativeaottests Debug`
results in, for each test,
```
    Starting:    nativeaot.SmokeTests.XUnitWrapper
      nativeaot/SmokeTests/Reflection/Reflection/Reflection.sh [FAIL]

        Return code:      1
        Raw output file:      /home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/nativeaot/SmokeTests/Reports/nativeaot.SmokeTests/Reflection/Reflection/Reflection.output.txt
        Raw output:
        BEGIN EXECUTION
        Running Native AOT:  /home/scott/github/runtimelab/dotnet.sh msbuild /p:IlcPath=/home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/Tests/Core_Root/nativeaot nativebuild.proj
        Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
        Copyright (C) Microsoft Corporation. All rights reserved.

          /usr/bin/clang
          Generating native code
          Unhandled exception. System.DllNotFoundException: Unable to load shared library 'objwriter' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libobjwriter: cannot open shared object file: No such file or directory
             at ILCompiler.DependencyAnalysis.ObjectWriter.InitObjWriter(String objectFilePath, String triple)
             at ILCompiler.DependencyAnalysis.ObjectWriter..ctor(String objectFilePath, NodeFactory factory) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:line 890
             at ILCompiler.DependencyAnalysis.ObjectWriter.EmitObject(String objectFilePath, IEnumerable`1 nodes, NodeFactory factory, IObjectDumper dumper) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs:line 974
             at ILCompiler.RyuJitCompilation.CompileInternal(String outputFile, ObjectDumper dumper) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilation.cs:line 61
             at ILCompiler.Compilation.ILCompiler.ICompilation.Compile(String outputFile, ObjectDumper dumper) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs:line 475
             at ILCompiler.Program.Run(String[] args) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler/Program.cs:line 734
             at ILCompiler.Program.Main(String[] args) in /home/scott/github/runtimelab/src/coreclr/tools/aot/ILCompiler/Program.cs:line 905
          Aborted
/home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/Tests/Core_Root/nativeaot/build/Microsoft.NETCore.Native.targets(240,5): error MSB3073: The command ""/home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/Tests/Core_Root/nativeaot/tools/ilc" @"/home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/nativeaot/SmokeTests/Reflection/Reflection/native/Reflection.ilc.rsp"" exited with code 134. [/home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/nativeaot/SmokeTests/Reflection/Reflection/nativebuild.proj] [/home/scott/github/runtimelab/src/tests/run.proj]
        Native AOT failed with exitcode: 1
```
It looks like the `DllImport` is not finding the `objwriter` which is on the filesystem in the same location as `ilc` as `libobjwriter.so`
My understanding is that `dlopen` will not look in the same directory as the running executable without some help.  `LD_LIBRARY_PATH` is the one way, but I've set that manually and it hasn't helped.  RPATH is the other way where the paths are in the executable but:
```
scott@DESKTOP-1VR0T42:~/github/runtimelab$ readelf -d /home/scott/github/runtimelab/artifacts/tests/coreclr/Linux.x64.Debug/Tests/Core_Root/nativeaot/tools/ilc

Dynamic section at offset 0x20888 contains 33 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN/netcoredeps]
```
`$ORIGIN`  on its own is not there, so it seems logical that it would fail, but how should this work as it clearly does in CI?  In the past there seems to have been some problems with the `lib` prefix, is that my problem?

**Comment by yowl** [comment]
Update: seems like its the `.so` extension that is the problem.  If I copy the lib to a file without the extension it finds it, so is there some way that the loader decides on the extension that might not be working right for me?

**Comment by jkotas** [comment]
I am not sure what may be causing the problems that you are seeing. I just run this on Ubuntu and it worked fine for me:

```
git clean -x -d -f .
./build.sh nativeaot+libs+nativeaot.packages -rc Debug -lc Release
cd src/tests
./build.sh nativeaot Debug /p:SmokeTestsOnly=true
./run.sh --runnativeaottests Debug
```

> My understanding is that dlopen will not look in the same directory as the running executable without some help.

The .NET native library loader does not pass the DllImport argument directly into `dlopen`. It creates full path and also tries different name variants by prepending `lib` and appending `.so` on Linux.

**Comment by yowl** [comment]
>     * Add an extra entry to the substitution file that hardcodes `GlobalizationMode.get_UseNls` to `false`

Why is this necessary when https://github.com/dotnet/runtimelab/blob/003167d8a7c5e5a990e96d759d32f5926eaf3885/src/coreclr/nativeaot/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs#L30 ?

**Comment by jkotas** [comment]
The property has to be inlined and optimizations have to kick in to get rid of the unreachable code. It is why you see the code in debug builds.

The alternatives for the substitution file would be:
- Change the property to constant ie `internal const bool UseNls = false;`. The problem with this is that Roslyn will produce a bunch of unreachable code warnings that are annoying to disable.
- Run IL linker while compiling System.Private.CoreLib, ie set `<ILLinkTrimAssembly>true</<ILLinkTrimAssembly>`. The problem with this is that ILLinker would trim a lot of other code that is needed by the runtime, and we would need linker annotations to keep it around.

**Comment by MichalStrehovsky** [comment]
The code in IL Linker can do this without the substitutions entry because it will just look at the method body and see that it's a constant: https://github.com/mono/linker/blob/6d05762d54a56309e49e6d8b70dca272e78e3e19/src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs#L66-L90 (see the ConstantExpressionMethodAnalyzer).

The equivalent code in the Native AOT compiler only looks at the substitutions file when eliminating unreachable basic blocks: https://github.com/dotnet/runtimelab/blob/50fa265894e2b360e29bd96472763d94ac1e217f/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs#L458-L473.

The main reason is perf - looking at the IL of called methods would make the compilation slower so I just scoped it out until we can really demonstrate it's needed. It would be nice here, but it's not essential.

**Comment by yowl** [comment]
@jkotas  @MichalStrehovsky  THanks, I understand that much better now.

**Comment by jkotas** [comment]
Fixed by #522
